src/lib/tls/asio/asio_stream.h

```diff
@@ -0,0 +1,487 @@
+         }
+
+      template<
+         typename VerifyCallback>
+      void set_verify_callback(VerifyCallback callback,
+                               boost::system::error_code& ec)
+         {
+         BOTAN_UNUSED(callback);
+         ec = make_error_code(Botan::TLS::error::not_implemented);
+         }
+
+      void set_verify_depth(int depth)
+         {
+         BOTAN_UNUSED(depth);
+         throw Not_Implemented("set_verify_depth is not implemented");
+         }
+
+      void set_verify_depth(int depth,
+                            boost::system::error_code& ec)
+         {
+         BOTAN_UNUSED(depth);
+         ec = make_error_code(Botan::TLS::error::not_implemented);
+         }
+
+      template <typename verify_mode>
+      void set_verify_mode(verify_mode v)
+         {
+         BOTAN_UNUSED(v);
+         throw Not_Implemented("set_verify_mode is not implemented");
+         }
+
+      template <typename verify_mode>
+      void set_verify_mode(verify_mode v,
+                           boost::system::error_code& ec)
+         {
+         BOTAN_UNUSED(v);
+         ec = make_error_code(Botan::TLS::error::not_implemented);
+         }
+
+      //
+      // -- -- handshake methods
+      //
+
+      void handshake()
+         {
+         boost::system::error_code ec;
+         handshake(ec);
+         boost::asio::detail::throw_error(ec, "handshake");
+         }
+
+      void handshake(boost::system::error_code& ec)
+         {
+         while(!native_handle()->is_active())
+            {
+            writePendingTlsData(ec);
+            if(ec)
+               {
+               return;
+               }
+
+            auto read_buffer = boost::asio::buffer(
+                                  this->m_core.input_buffer,
+                                  m_nextLayer.read_some(this->m_core.input_buffer, ec));
+            if(ec)
+               {
+               return;
+               }
+
+            try
+               {
+               native_handle()->received_data(static_cast<const uint8_t*>(read_buffer.data()),
+                                              read_buffer.size());
+               }
+            catch(...)
+               {
+               ec = Botan::TLS::convertException();
+               return;
+               }
+
+            writePendingTlsData(ec);
+            }
+         }
+
+      template <typename HandshakeHandler>
+      BOOST_ASIO_INITFN_RESULT_TYPE(HandshakeHandler,
+                                    void(boost::system::error_code))
+      async_handshake(HandshakeHandler&& handler)
+         {
+         BOOST_ASIO_HANDSHAKE_HANDLER_CHECK(HandshakeHandler, handler) type_check;
+
+         boost::asio::async_completion<HandshakeHandler,
+               void(boost::system::error_code)>
+               init(handler);
+
+         auto op = create_async_handshake_op(std::move(init.completion_handler));
+         op(boost::system::error_code{}, 0, 1);
```

Comment:
The system of launching composed operations is different now (and a lot better). See: https://github.com/boostorg/asio/commit/0568d3bf0b492887c6c48d05f5492a782dafa746

Beast supports all these as of v217